### PR TITLE
net/udpbroadcastrelay plugin - Initial commit

### PR DESCRIPTION
**Pull request by marjohn56**
udpbroadcastrelay - a plugin to control the daemon of the same name. The original daemon was called udp-broadcast-relay-redux, that daemon has been modified to better suit our purposes and is ready to be uploaded to wherever I am told to send it.

This plugin handles all the configurations required for multiple instances and displays those instances status in the services status UIs.

Usage is simple, port, Instance ID - which is also used as part of the pid, multicast address if used and drop down select of the interfaces to use. Source address is also an option for the likes of Chromecast.

The daemon is included in this commit for testing purposes only.

### CODE REVIEW

File: net/udpbroadcastrelay/Makefile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+PLUGIN_NAME=		os-udpbroadcastrelay
+PLUGIN_VERSION=		0.1
+PLUGIN_REVISION=	0.1
+PLUGIN_DEV=         YES
```
Comment (mimugmail) [inline net/udpbroadcastrelay/Makefile]: PLUGIN_DEVEL

File: net/udpbroadcastrelay/pkg-descr [hunk 2]
```diff
@@ -0,0 +1,29 @@
+udbproadcastrelay is a USB multicast relayer. Its intended use is to
```
Comment (mimugmail) [inline net/udpbroadcastrelay/pkg-descr]: USB? :)

File: net/udpbroadcastrelay/src/opnsense/mvc/app/controllers/OPNsense/UDPBroadcastRelay/Api/ServiceController.php [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+use \OPNsense\Core\Backend;
+
+/**
+ * Class ServiceController Handles settings related API actions for the UDPBroadcastRelay
+ * @package OPNsense\UDPBroadcastRelay
+ */
+class ServiceController extends ApiMutableModelControllerBase
+{
+    protected static $internalModelName = 'ts';
+    protected static $internalModelClass = '\OPNsense\\UDPBroadcastRelay\\UDPBroadcastRelay';
```
Comment (mimugmail) [inline net/udpbroadcastrelay/src/opnsense/mvc/app/controllers/OPNsense/UDPBroadcastRelay/Api/ServiceController.php]: Is there a reason for double backspace?

File: net/udpbroadcastrelay/src/opnsense/mvc/app/controllers/OPNsense/UDPBroadcastRelay/Api/SettingsController.php [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+use \OPNsense\UDPBroadcastRelay\UDPBroadcastRelay;
+use \OPNsense\Base\UIModelGrid;
+/**
+ * Class SettingsController Handles settings related API actions for the UDPBroadcastRelay
+ * @package OPNsense\UDPBroadcastRelay
+ */
+class SettingsController extends ApiMutableModelControllerBase
+{
+    protected static $internalModelName = 'ts';
+    protected static $internalModelClass = '\OPNsense\TrafficShaper\TrafficShaper';
```
Comment (mimugmail) [inline net/udpbroadcastrelay/src/opnsense/mvc/app/controllers/OPNsense/UDPBroadcastRelay/Api/SettingsController.php]: TrafficShaper?

File: net/udpbroadcastrelay/src/opnsense/mvc/app/controllers/OPNsense/UDPBroadcastRelay/ItemController.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace OPNsense\UDPBroadcastRelay;
+
+/**
+ * Class ItemController
+ * @package OPNsense\UDPBroadcastRelay
+ */
+class ItemController extends \OPNsense\Base\IndexController
```
Comment (mimugmail) [inline net/udpbroadcastrelay/src/opnsense/mvc/app/controllers/OPNsense/UDPBroadcastRelay/ItemController.php]: Where is ItemController used? Didn't fint it. Maybe you can delete this file (after testing)

File: net/udpbroadcastrelay/src/opnsense/scripts/OPNsense/UDPBroadcastRelay/setup.sh [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+chmod +x /usr/local/etc/rc.d/os-udpbroadcastrelay
```
Comment (mimugmail) [inline net/udpbroadcastrelay/src/opnsense/scripts/OPNsense/UDPBroadcastRelay/setup.sh]: You can remove this file, @fichtner will do a chmod in git :)

File: net/udpbroadcastrelay/src/opnsense/service/conf/actions.d/actions_udpbroadcastrelay.conf [hunk 7]
```diff
@@ -0,0 +1,30 @@
+[start]
+command:
+    /usr/local/opnsense/scripts/OPNsense/UDPBroadcastRelay/setup.sh;
```
Comment (mimugmail) [inline net/udpbroadcastrelay/src/opnsense/service/conf/actions.d/actions_udpbroadcastrelay.conf]: Also remove here

File: net/udpbroadcastrelay/src/opnsense/service/conf/actions.d/actions_udpbroadcastrelay.conf [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    /usr/local/etc/rc.d/os-udpbroadcastrelay start
+parameters:%s
```
Comment (mimugmail) [inline net/udpbroadcastrelay/src/opnsense/service/conf/actions.d/actions_udpbroadcastrelay.conf]: Does start/stop/etc really expect a parameter?

**Comment by marjohn56** [comment]
OK, thanks for the quick review. It appears I pushed a slightly earlier version.:) This is due to me using Github Desktop on my PC and transferring between the instance running on my emulator.. I know, bound to lead to mistakes. Update shortly.

File: net/udpbroadcastrelay/src/opnsense/service/conf/actions.d/actions_udpbroadcastrelay.conf (hunk 8 above)
Comment (marjohn56) [reply net/udpbroadcastrelay/src/opnsense/service/conf/actions.d/actions_udpbroadcastrelay.conf]: Erm.. I thought it did, apparently it doesn't - removed.

File: net/udpbroadcastrelay/src/opnsense/mvc/app/controllers/OPNsense/UDPBroadcastRelay/ItemController.php (hunk 5 above)
Comment (marjohn56) [reply net/udpbroadcastrelay/src/opnsense/mvc/app/controllers/OPNsense/UDPBroadcastRelay/ItemController.php]: It was there in the 'borrowed' source. I expect it's part of a template so I left it for future projects.

> I only had a quick view, some things can be deleted, or blindly copied. You put much effort in it, really cool! :) Can you also please add your copyright to every file? I know you just copied it, but that's ok. Every php and volt needs one. Binary pull is dirty, can you post your repo and maybe try to add port to opnsense ports?

I can put it there, with a link to this. @fichtner will need to do his magic on it though as it's a very basic makefile etc.

**Comment by marjohn56** [comment]
Added daemon commit to ports.
https://github.com/opnsense/ports/pull/84

**Comment by marjohn56** [comment]
OK, Kev found an issue I had not noticed. When the system is rebooted if there are two instances one does not start, and after checking I found that when started with configctl it fails, it works fine if started from the shell with service "os-udpbroadcastrelay start". The further I looked into this it appears where a check is made for the pid in rc.d/os-udpbroadcastrelay at line 35 it fails when called from configctl. I can remove the check and it works, but why does it fail?

File: net/udpbroadcastrelay/Makefile [hunk 9]
```diff
@@ -0,0 +1,8 @@
+PLUGIN_NAME=		os-udpbroadcastrelay
+PLUGIN_VERSION=		0.5
+PLUGIN_REVISION=	0.0
```
Comment (mimugmail) [inline net/udpbroadcastrelay/Makefile]: you can remove this line as long as it's in devel

**Comment by mimugmail** [comment]
Have a look at the console if it's startup related

**Comment by marjohn56** [comment]
Fixed it, this works.
`cmd_string=`basename ${procname:-${command}}`
      cmd="ps ax -o pid= -o command= | grep $cmd_string | grep -e "$osudpbroadcastrelay_flags" | grep -v grep | awk '{ print $1 }"


      ps_pid=$cmd`

File: net/udpbroadcastrelay/src/opnsense/service/conf/actions.d/actions_udpbroadcastrelay.conf (hunk 8 above)
Comment (marjohn56) [reply net/udpbroadcastrelay/src/opnsense/service/conf/actions.d/actions_udpbroadcastrelay.conf]: Yes it does for services. Along with another fix in inc/plugins.inc.d the items in the services widget now start and stop independently.

**Comment by marjohn56** [comment]
Added multiple multicast address options and use original TTL as Identifier for echoed packets if needed. Now default uses new DSCP to identify the packets and leaves TTL untouched.

**Comment by nivek1612** [comment]
@mimugmail do we have any timelines as to when this may make production. Really useful plugin for those of us using vlans.

**Comment by mimugmail** [comment]
@nivek1612 there are some things in this PR still unclear (ts, double back space, binary blob, copyrights)
@marjohn56 I'm sure I already asked in IRC but cant find it. Where on github is the source you are using in the ports PR? Adding code to ports tree directly is not allowed, only patches. I can try to build a correct makefile and maybe also upstream to FreeBSD itself

**Comment by marjohn56** [comment]
https://github.com/marjohn56/ports/tree/udpbroadcastrelay

**Comment by mimugmail** [comment]
@marjohn56 can you create a new own repo and upload the code?
There should be no relation to the whole ports tree.

**Comment by marjohn56** [comment]
https://github.com/marjohn56/UDPBroadcastRelay

**Comment by mimugmail** [comment]
I don't want to be picky or offend, but:

- can you please make it lowercase only
- tag a release number
- check if the previous license was bsdclause-2 and if not, get a permission to change
- add this to your makefile

```
all:
        $(CC) $(CFLAGS) -g main.c -o udpbroadcastrelay
        cp udpbroadcastrelay /usr/local/bin/
        chmod 755 /usr/local/bin/udpbroadcastrelay
```

**Comment by marjohn56** [comment]
OK, done...

https://github.com/marjohn56/udpbroadcastrelay

Just one thing, the scripts call sbin/udpbroadcastrelay not /bin, so it needs changing here or there... I care not which.

Cheers...

**Comment by mimugmail** [comment]
Can you recheck? I can't find the repo. If sbin is ok then just place it in the Makefile, it was just a guess from mine. Please keep this PR open and only remove the binary :)

**Comment by marjohn56** [comment]
It's there. I've deleted the original repo and the mixed case one and created a new one at this link:
https://github.com/marjohn56/udpbroadcastrelay
I posted it above.

**Comment by mimugmail** [comment]
There's a 404 when I click the link and it's not listed in your repo's list.

**Comment by marjohn56** [comment]
Try a cache flush. ;)

Alternatively, go here 

https://github.com/marjohn56/udpbroadcastrelay/commit/5e07941f44872282b20b792f72417d83e6a339b6

**Comment by marjohn56** [comment]
Try it now, for some reason it got set to Private.

https://github.com/marjohn56/udpbroadcastrelay

**Comment by mimugmail** [comment]
Thanks, looks better now. 
Can you please tag a release?

**Comment by marjohn56** [comment]
Not sure what you mean, I could of course just be being thick. You want a version in the source itself?

**Comment by mimugmail** [comment]
In Github you can tag a release, this will be the source for FreeBSD ports system to manage the rest:

https://github.com/opnsense/core/releases

Just tag version 1.0 and I'll do the rest :)

**Comment by marjohn56** [comment]
Well you learn something new every day…

Have done so.

**Comment by mimugmail** [comment]
Thx, can you merge my PR in your repo so I can go on? :)

**Comment by marjohn56** [comment]
😊 Is done.

**Comment by fichtner** [comment]
@marjohn56 @mimugmail udpbroadcastrelay is in packages come 20.1.3 so do we want this in as well or is it still WIP?

**Comment by marjohn56** [comment]
No, it's working nicely. Run it up and try it.

**Comment by mimugmail** [comment]
It would be nice to have a devel pkg .. so, please merge. 
It needs cleanup for sure ... but easier to work on when in master.

**Comment by marjohn56** [comment]
I'd love to know what this cleanup is... pray tell. :)

**Comment by mimugmail** [comment]
Makefile needs a PLUGIN_DEPENDS with udpbroadcastrelay (before merge), rc file doesn't really look like a rc file but not really important, Model name "ts" and double back ticks in ServiceController, ItemController seems unused, Copyright in Model, Model mount differs from ACL, volt has stuff from FTP Proxy. The first is a show stopper, rest can be done after merge

**Comment by marjohn56** [comment]
I merged some of your stuff last week, the only thing I did not merge was the params, which are needed.

**Comment by marjohn56** [comment]
Just tested the source on Github and it no longer works, reverted to my pkg source that's on the forum, works fine. I'll take a look later and see what's causing the problem.

**Comment by mimugmail** [comment]
It might be the missing setup.sh which sets the chmod, but this will be done by @fichtner since you can set these chmods in git directly

**Comment by mimugmail** [comment]
(and of course the removed binary)

**Comment by marjohn56** [comment]
I'll check later, off to do some proper work for a few hours.

File: net/udpbroadcastrelay/Makefile [hunk 10]
```diff
@@ -1,5 +1,5 @@
 PLUGIN_NAME=		os-udpbroadcastrelay
-PLUGIN_VERSION=		0.6
+PLUGIN_VERSION=		0.7
 PLUGIN_DEVEL=         YES
```
Comment (mimugmail) [inline net/udpbroadcastrelay/Makefile]: You need 
`PLUGIN_DEPENDS= udpbroadcastrelay`

Otherwise it will not be installed when installing the plugin

Comment (marjohn56) [reply net/udpbroadcastrelay/Makefile]: Indeed, my commit to add that went out seconds before your comment :)

Squashed into one commit now.

**Comment by mimugmail** [comment]
What happens on your end when you remove ItemController.php? Does everything still work? I quite sure it's not needed.

**Comment by marjohn56** [comment]
My router just exploded!!

Nah.. nothing at all, we'll remove it.

**Comment by marjohn56** [comment]
Removed.

**Comment by mimugmail** [comment]
Where in the volt do you handle the restart action when changing anything? May I'm blind? Does the daemon restart if you change things?

**Comment by marjohn56** [comment]
Yes it does restart, if you have multiple instances they will restart or stop too. Fire it up and add a couple of relays is they way to see whats happening, if you only have one LAN then create a a VLAN or two and add a couple of instances on different ports or multicast addresses.

Capture1
Capture2

**Comment by marjohn56** [comment]
Think it may be time to close this PR and issue a new clean one, far too much traffic here I think.

**Comment by fichtner** [comment]
I don‘t mind, will squash at the end.

**Comment by marjohn56** [comment]
OK.. I was doing it to make it easier to see what was going on, but if you are happy to leave it be then so am I. :)

**Comment by mimugmail** [comment]
What's the current status here? The binary is already available with 20.1.3. 
If we merge we could have a devel pkg in 20.1.4 to go out for testing :)

**Comment by marjohn56** [comment]
bump... anything happening with this or shall I just squash it to make it quicker to pull?

**Comment by mimugmail** [comment]
I installed your branch locally here. After adding one instance I got this in configd.log:

```
Apr 15 14:11:58 OPNsense configd.py: [692e0c30-c724-46a4-b960-aeee1105d1f2] get udpbroadcastrelay instance status
Apr 15 14:11:58 OPNsense configd.py: [692e0c30-c724-46a4-b960-aeee1105d1f2] returned exit status 126

```

Then I set chmod 755 to rc sciript (which @fichtner will take care of after merge) and then I got:

```
Apr 15 14:12:49 OPNsense configd.py: [107f3740-9299-448a-89e7-1a84777c681a] get udpbroadcastrelay instance status
Apr 15 14:12:49 OPNsense configd.py: [107f3740-9299-448a-89e7-1a84777c681a] returned exit status 1

```

I could only start it via dashboard -> services -> UDP Broadcast Relay (which needs description).
Then everything seems to work.

Volt looks a bit different than the ususal plugins but it works :)

@fichtner for devel pkg for me it's ok :)

**Comment by fichtner** [comment]
Merged, thanks. Some cleanups follow.

**Comment by marjohn56** [comment]
Yes. it needs the kick from the services start for the first time. There is a doc for it, I think it's already been PR'd

**Pull request merged**